programs/stake_api/src/stake_instruction.rs

```diff
@@ -10,26 +10,84 @@ use solana_sdk::system_instruction;
 
 #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
 pub enum StakeInstruction {
+    /// Initialize the stake account as a Delegate account.
+    ///
+    /// Expects 2 Accounts:
+    ///    0 - payer (TODO unused/remove)
+    ///    1 - Delegate StakeAccount to be initialized
+    InitializeDelegate,
+
+    // Initialize the stake account as a MiningPool account
+    ///
+    /// Expects 2 Accounts:
+    ///    0 - payer (TODO unused/remove)
+    ///    1 - MiningPool StakeAccount to be initialized
+    InitializeMiningPool,
+
     /// `Delegate` or `Assign` a stake account to a particular node
-    ///  expects 2 KeyedAccounts:
-    ///     StakeAccount to be updated
-    ///     VoteAccount to which this Stake will be delegated
+    ///
+    /// Expects 3 Accounts:
+    ///    0 - payer (TODO unused/remove)
+    ///    1 - Delegate StakeAccount to be updated
+    ///    2 - VoteAccount to which this Stake will be delegated
     DelegateStake,
 
     /// Redeem credits in the stake account
-    ///  expects 3 KeyedAccounts: the StakeAccount to be updated
-    ///  and the VoteAccount to which this Stake will be delegated
+    ///
+    /// Expects 4 Accounts:
+    ///    0 - payer (TODO unused/remove)
+    ///    1 - MiningPool Stake Account to redeem credits from
+    ///    2 - Delegate StakeAccount to be updated
+    ///    3 - VoteAccount to which the Stake was delegated to previously
     RedeemVoteCredits,
 }
 
-pub fn create_account(from_id: &Pubkey, staker_id: &Pubkey, lamports: u64) -> Instruction {
-    system_instruction::create_account(
-        from_id,
-        staker_id,
-        lamports,
-        std::mem::size_of::<StakeState>() as u64,
-        &id(),
-    )
+pub fn create_delegate_account(
+    from_id: &Pubkey,
+    staker_id: &Pubkey,
+    lamports: u64,
+) -> Vec<Instruction> {
+    vec![
+        system_instruction::create_account(
+            from_id,
+            staker_id,
+            lamports,
+            std::mem::size_of::<StakeState>() as u64,
+            &id(),
+        ),
+        Instruction::new(
+            id(),
+            &StakeInstruction::InitializeDelegate,
+            vec![
+                AccountMeta::new(*from_id, true),
+                AccountMeta::new(*staker_id, false),
+            ],
+        ),
+    ]
+}
+
+pub fn create_mining_pool_account(
+    from_id: &Pubkey,
+    staker_id: &Pubkey,
+    lamports: u64,
+) -> Vec<Instruction> {
+    vec![
+        system_instruction::create_account(
+            from_id,
+            staker_id,
+            lamports,
+            std::mem::size_of::<StakeState>() as u64,
+            &id(),
+        ),
+        Instruction::new(
```

Comment:
I liked that StakeState::default() was a delegate to Pubkey::default()...  Was there a complication with that?

Reply:
Yeah.  Because the StakeState represents two different things, Delegate or MiningPool, that enum needs to be set once upon account creation before it knows what it is.   Before we defaulted to a Delegate, and there was no way to turn it into a MiningPool account.

Reply:
nevermind, I get it.  is cleaner/better to test for an enum than for delegate == Pubkey::default() when figuring a type

Reply:
we can re-cast anytime we want, but my "solution" was ugly

Reply:
it occurs to me that there was no way to construct a miningpool, and that waiting until a redemption request was made to determine account type opens a hole: a staking account whose delegate hasn't been initialized could be hijacked to be a miningpool

not just ugly: a bug